util/src/main/java/io/kubernetes/client/util/Cache.java

```diff
@@ -0,0 +1,269 @@
+  private Function<CallGeneratorParams, Call> callGenerator;
+  private ApiClient client;
+  private Watch<ApiType> watch;
+  private Type watchType;
+  private ApiListType list;
+  private Class<?> listClass;
+  private Thread watchThread;
+  private Integer timeout;
+  private Map<String, List<ApiType>> namespaceIndex;
+
+  private static final Logger log = LoggerFactory.getLogger(Cache.class);
+
+  private Cache() {}
+
+  public Cache(
+      ApiClient client,
+      Function<CallGeneratorParams, Call> callGenerator,
+      Class<?> listClass,
+      Type watchType) {
+    this.client = client;
+    // TODO: Parameter?
+    this.timeout = 5 * 60;
+    this.callGenerator = callGenerator;
+    this.watch = null;
+    this.watchType = watchType;
+    this.listClass = listClass;
+    this.list = null;
+    this.namespaceIndex = new HashMap<>();
+  }
+
+  private void init() throws ApiException, IOException {
+    if (list == null) {
+      updateList();
+    }
+    if (watch == null) {
+      startWatch();
+    }
+  }
+
+  public synchronized ApiListType list() throws ApiException, IOException {
+    init();
+    return list;
+  }
+
+  public synchronized List<ApiType> listForNamespace(String namespace)
+      throws ApiException, IOException {
+    init();
+    return namespaceIndex.get(namespace);
+  }
+
+  public synchronized ApiType get(String namespace, String name) throws ApiException, IOException {
+    init();
+    return getItem(namespace, name);
+  }
+
+  public void close() throws IOException {
+    if (watch != null) {
+      watch.close();
+    }
+  }
+
+  private Call makeCall(Boolean watch) {
+    CallGeneratorParams params = new CallGeneratorParams(watch, null, timeout);
+    return callGenerator.apply(params);
+  }
+
+  private void updateList() throws ApiException, IOException {
+    Call call = makeCall(Boolean.FALSE);
+    com.squareup.okhttp.Response response = call.execute();
+    if (!response.isSuccessful()) {
+      String respBody = null;
+      try (ResponseBody body = response.body()) {
+        if (body != null) {
+          respBody = response.body().string();
+        }
+      } catch (IOException e) {
+        throw new ApiException(
+            response.message(), e, response.code(), response.headers().toMultimap());
+      }
+      throw new ApiException(
+          response.message(), response.code(), response.headers().toMultimap(), respBody);
+    }
+    list =
+        (ApiListType) client.getJSON().getGson().fromJson(response.body().charStream(), listClass);
+    try {
+      List<ApiType> itemList = getItemList();
+      for (ApiType item : itemList) {
+        indexItem(item);
+      }
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException ex) {
+      throw new ApiException(ex);
+    }
+  }
+
+  private ApiType getItem(String namespace, String name) {
+    try {
+      Method itemsField = listClass.getDeclaredMethod("getItems");
```

Comment:
Reflection is slow, we should consider doing this abstraction only when necessary.

Reply:
indeed, any better way to extract the list other than reflection?